lib/cartopy/geodesic/_geodesic.pyx

```diff
@@ -0,0 +1,253 @@
+    """
+    Defines an ellipsoid on which to solve geodesic problems. 
+
+    """
+    cdef geod_geodesic* geod
+
+    def __cinit__(self, radius=6378137, flattening=1/298.257223563):
+        """
+        Create an ellipsoid with a given radius and flattening.
+
+        Kwargs:
+
+            * radius  - Equatorial radius (metres).
+
+            * flattening - Flattening of ellipsoid.
+                           Setting flattening = 0 gives a sphere. Negative 
+                           flattening gives a prolate ellipsoid. If 
+                           flattening > 1, set flattening to 1/flattening.
+
+        """
+        # allocate some memory (filled with random data)
+        self.geod = <geod_geodesic*> PyMem_Malloc(sizeof(geod_geodesic))
+        if not self.geod:
+            raise MemoryError()
+        geod_init(self.geod, radius, flattening)
+
+    def direct(self, points, azimuths, distances):
+        """
+        Solve the direct geodesic problem where the length of the geodesic is 
+        specified in terms of distance.
+
+        Can accept and broadcast length 1 arguments. For example, given a single
+        start point and distance, an array of different azimuths can be supplied
+        to locate multiple endpoints.
+
+        Args:
+
+            * points - An n (or 1) by 2 numpy.ndarray, list or tuple of lon-lat
+                       points.
+                       The starting point(s) from which to travel.
+
+            * azimuths - A length n (or 1) numpy.ndarray or list of azimuth 
+                         values (degrees).
+
+            * distances - A length n (or 1) numpy.ndarray or list of distances 
+                          values (metres).
+
+        Returns:
+            An n by 3 np.ndarray of lons, lats, and azimuths of the located 
+            endpoint. 
+
+        """
+
+        cdef int n_points, i
+        cdef double[:, :] pts
+        cdef double[:] azims, dists
+
+        # Create numpy arrays from inputs, and ensure correct shape. Note: 
+        # reshape(-1) returns a 1D array from a 0 dimensional array as required 
+        # for broadcasting.
+        pts = np.array(points, dtype = np.float64).reshape((-1, 2))
+        azims = np.array(azimuths, dtype = np.float64).reshape(-1)
+        dists = np.array(distances, dtype = np.float64).reshape(-1)
+
+        n_points = max(pts.shape[0], azims.size, dists.size)
+
+        # Broadcast any length 1 arrays to the correct size.
+        try:
+            tmp = np.zeros((n_points, 2))
+            tmp[:, 0] += pts[:, 0]
+            tmp[:, 1] += pts[:, 1]
+
+            pts = tmp
+
+            azims = np.zeros(n_points) + azims[:]
+
+            dists = np.zeros(n_points) + dists[:]
+
+        except ValueError:
+            raise ValueError("Inputs must have common length n or length one.")
+
+        cdef double[:, :] return_pts = np.empty((n_points, 3))
+        cdef double[:] lon, lat, azi
+
+        lon = np.empty(n_points)
```

Comment:
The proj4 functions we're wrapping return values by changing the `lon`, `lat` and `azi` variables that we've passed pointers to, rather than returning a value directly and so we need the variables to get the result and copy it into `return_pts`.